### PR TITLE
feat(plans): per-plan health-score badge (closes #376 scope)

### PR DESCRIPTION
**Pull request by cristim**
## Summary

Revives the per-plan health-score badge from PR #376 (closed stale,
never merged), implemented fresh against current `main`. Adds a 0-100
"Plan health" badge to every purchase plan card on the Plans tab.
Green >= 80, amber 50-79, red < 50. The badge's tooltip enumerates
exactly which factors subtracted from 100 so a bad score is actionable
instead of opaque.

Historical context: issue #340 (closed, unrelated broader UI-revamp
scope) is where PR #376's body originally noted this as a deferred
follow-up idea. This PR does not close either #340 or #376 -- #340 is
out of scope and already closed, and #376 is a PR, not an issue.

## Scoring

Pure backend function `computePlanHealth` (`internal/api/plan_health.go`).
Starts at 100, subtracts weighted penalties, clamps to [0, 100]:

| Factor                  | Weight  | Trigger                                                            |
|-------------------------|---------|--------------------------------------------------------------------|
| `overdue`               | -30     | enabled AND `next_execution_date` is in the past                    |
| `failed_executions`     | -10 x n | n failed executions in the lookback window (n capped at 4 -> max -40) |
| `canceled_executions`   | -5 x n  | n canceled executions in the lookback window (n capped at 4 -> max -20) |
| `stalled`               | -15     | enabled, past the first interval, but `current_step == 0`           |
| `behind_schedule`       | -20     | `current_step` lags the step implied by start_date + interval       |
| `disabled_midway`       | -25     | disabled with `0 < current_step < total_steps`                     |

A completed plan (`RampSchedule.IsComplete()`, guarded against the
degenerate `total_steps == 0` case) short-circuits to 100: post-completion
failure rows and a toggle-off are expected end-of-life noise, not
something an operator should chase.

Scoring details worth calling out:

- **Lookback window.** The two execution-count factors count only rows
  whose `updated_at` falls within the trailing `planHealthLookbackDays`,
  which is deliberately the same constant as
  `config.DefaultExecutionTTLDays` (30 days) that the execution-retention
  sweep uses. That sweep deletes canceled rows past the horizon but never
  failed ones, so counting over any longer span would measure the two
  factors over different periods and permanently penalize a plan for
  failures years ago that no number of clean runs could clear. Every
  factor note names the window, so the number on screen is unambiguous.
- **`stalled` and `behind_schedule` are mutually exclusive** -- at most
  one of the pair ever applies.
- **Schedule factors are skipped** when `step_interval_days <= 0`
  (immediate ramps have no schedule position to fall behind on, and this
  is the only divisor) or when `start_date` is zero (measuring elapsed
  time from a zero `time.Time` would fabricate a ~740000-day penalty out
  of a missing input; `RampSchedule.GetNextPurchaseDate` guards the same
  field the same way).
- **Canceled counts both spellings** (`config.StatusCanceled` and
  `config.LegacyStatusCanceled`) so pre-#1278 rows are not invisible to
  scoring.

## Implementation notes

- **New store method, aggregated in SQL.**
  `config.StoreInterface.CountExecutionsByPlanAndStatus(ctx, statuses, since)`
  returns exact per-plan counts via `GROUP BY plan_id, status`. It is
  deliberately NOT derived from `GetExecutionsByStatuses`: that method is
  capped by `limit` across ALL plans, so counting per-plan rows out of it
  silently understates any plan whose executions fall outside the newest
  page -- rendering a confidently-healthy badge for an unhealthy plan. The
  result set is bounded by plans x statuses, not by execution volume.
- **Windowed on `updated_at`, not `scheduled_date`.** Executions are
  created up front for the whole ramp, so a pending row carries a
  `scheduled_date` months in the future, and `CancelExecutionAtomic`
  leaves that date untouched. Windowing on `scheduled_date` would count a
  purchase canceled today under a date next year. `updated_at` is stamped
  by `CancelExecutionAtomic` / `TransitionExecutionStatus` and backstopped
  by the `update_purchase_executions_updated_at` trigger, so it is when
  the row actually entered the counted status.
- **NULL `plan_id` rows excluded.** `plan_id` has been nullable since
  migration 000033 (direct-execute purchases from the Recommendations
  page have no originating plan; deleting a plan SET NULLs its
  executions). Those rows belong to no plan and are filtered in SQL; the
  scan still goes through `sql.NullString` so a NULL can never turn into a
  scan error that blanks the score for every plan.
- **Failure mode is an explicit unknown, never a fabricated score.** If
  the counts fetch fails, `health_score` is left nil, serializes as
  `"health_score": null`, and renders as an explicit "unknown" badge. A
  score is a number operators make purchasing decisions from, so an
  uncomputable one must stay absent -- substituting 100 ("healthy") or 0
  ("broken") would be indistinguishable from a real value. The request
  itself still succeeds; the Plans page is not primarily about execution
  history, so a 500 would be the worse trade.
- **New `PlanWithHealth` response envelope** wraps `config.PurchasePlan`
  with `health_score` (nullable) / `health_factors`, so neither field
  persists into the DB-backed struct.
- **Frontend distinguishes three states**: field absent (older cached API
  response) -> no badge at all; `null` or non-finite -> explicit unknown
  badge; a number -> colored badge.
- The tooltip (native `title` attribute) is built from
  `escapeHtmlAttr`-escaped factor notes, per this repo's
  all-API-fields-through-an-escaper convention.
- Reuses the existing `status-badge`/`badge-success`/`badge-warning`/
  `badge-danger` CSS classes already used by the Overdue badge on the
  same card -- no new CSS needed.

## Test plan

Verified on the rebased head (branch rebased onto `main` to pick up the
govulncheck fix for GO-2026-6061, grpc v1.80.0 -> v1.82.1; the rebase was
clean, no conflicts, and touched no dependency files):

- [x] `gofmt -l .` (exit 0, no output)
- [x] `go vet ./...` (exit 0)
- [x] `go build ./...` (exit 0)
- [x] `go test ./internal/api/... ./internal/config/... ./internal/mocks/...
      ./internal/server/... ./internal/analytics/... -count=1` (exit 0)
- [x] `gocyclo -over 10` on every touched non-test Go file (exit 0, no
      violations -- this repo's pre-commit gate is stricter than
      golangci's min-complexity 15)
- [x] `cd frontend && npx tsc --noEmit` (exit 0)
- [x] `cd frontend && npx jest` (exit 0; 86/86 suites, 2736 passed, 1
      skipped). Note: under a `de_DE` shell locale, 8 currency-formatting
      tests in `utils.test.ts` / `approval-details.test.ts` /
      `riexchange.test.ts` fail on the thousands separator (`$1.000` vs
      `$1,000`). That is an `Intl.NumberFormat` locale artifact in files
      this branch does not touch; under `LC_ALL=en_US.UTF-8` (what CI
      runs) the full suite is green.

## Retention / health-window fixes (d69ed8fd, 9166bc5d)

The score counts execution rows over a 30-day window on `updated_at`, so the
retention sweep must not delete a row while that window still covers it. It did.

**`CleanupOldExecutions` purged rows inside their own health window.** It
retained canceled rows on `scheduled_date`, while `CountExecutionsByPlanAndStatus`
counts them on `updated_at`, and `CancelExecutionAtomic` stamps `updated_at`
while leaving `scheduled_date` untouched. A plan's executions are created up
front and `parseCreatePurchasesRequest` validates only the date *format*, not
that the start date is in the future, so a purchase canceled today can carry a
40-day-old `scheduled_date`. The next sweep deleted it while the score was still
counting it, and the plan's badge gained up to 20 points overnight with nothing
having changed about the plan.

The same purge was reachable through the `expires_at` branch, and not only for
canceled: the score also counts `failed`, at -10 each up to -40. `expires_at` is
written once at insert and never rewritten when a row later reaches a terminal
status, so a row that failed or was canceled today can carry a lapsed
`expires_at` and be deleted by that branch on the same sweep.

The predicate is now three fully parenthesized branches: `completed` retains on
`scheduled_date`, the canceled spellings retain on `updated_at`, and the
`expires_at` branch excludes every status the score counts. That exclusion is
parameterized from `config.HealthScoredExecutionStatuses`, the same slice
`plan_health.go` passes to `CountExecutionsByPlanAndStatus`, so a status added
to the score extends the sweep's exclusion in the same edit instead of drifting
from a second literal list.

**`behind_schedule` quoted a step the plan does not have.** `expectedStep` was
never clamped to `TotalSteps`, so a 4-step weekly ramp started 90 days ago
rendered "on step 2, expected step 12 by now". The -20 penalty was correct; the
number the operator was asked to reconcile against was fabricated by the
arithmetic.

### Reachability of the `expires_at` arm - correcting an earlier claim

An earlier version of this description presented the `expires_at` arm as a live
defect. That overstates it. `purchase_executions.expires_at` has exactly one
writer, `SavePurchaseExecution` via `timeFromTTL(execution.TTL)`, and no
production path assigns `PurchaseExecution.TTL`, so the column is NULL on every
row current code creates and the branch is inert for them. The exclusion is
**defence in depth**, not a fix for a scenario reachable today.

Whether legacy DynamoDB-era imported rows carry a non-NULL `expires_at` is an
open question that needs a production query; it cannot be settled from code, and
this PR asserts nothing either way. Note also that `expires_at` is the row's own
TTL column, *not* the approval-token deadline - that is `approval_token_expires_at`,
a separate column added by migration 000051. The doc comment said otherwise and
has been corrected, along with a stale claim that no code writes an `'expired'`
status (migration 000089's CHECK lists it, and `expireStaleExecutionsSweep`
writes it).

### Behaviour change worth knowing about

A canceled row with a **future** `scheduled_date` was previously retained
forever, because `scheduled_date < NOW() - retention` never became true for it.
It now leaves History 30 days after the cancellation rather than 30 days after
the date it was scheduled for, so a purchase canceled today for a 2-year-out
date disappears in a month instead of in two years. The direction is right - it
closes an unbounded-retention leak and matches what the health window already
assumed - but it is user-visible rather than a pure bug fix.

`failed` rows are deliberately still never deleted by any branch, which is what
`plan_health.go`'s lookback comment already claimed and the score's window
depends on.

### Verifying the retention fix

The behavioural proof is an integration test against a real PostgreSQL, because
the defect is in what the predicate *means*, not how it is spelled:

```bash
go test -tags=integration ./internal/config/ \
  -run TestPostgresStoreDB_CleanupOldExecutions_RetainsCanceledInsideHealthWindow
```

**`-tags=integration` is load-bearing.** Without it the file is excluded by its
build tag, the run reports "no tests to run" and exits 0, and that false green is
indistinguishable from a pass.

It seeds nine rows across the status x timestamp matrix and asserts exactly which
survive a 30-day sweep, in both directions. Confirmed failing before each fix:
pre-`d69ed8fd` the database deleted all three rows that must be retained;
pre-`9166bc5d` it deleted the `failed`-today row, which the suite had no case for
at all. It also asserts a genuinely old canceled row IS still deleted and that
`RowsAffected` matches, so neither fix can pass by quietly disabling retention.

Unit-level guards: `TestPGXMock_CleanupOldExecutions_RetainsCanceledOnUpdatedAt`
pins each branch's retention column, and
`TestPGXMock_CleanupOldExecutions_ExcludesEveryHealthScoredStatus` pins the `$2`
argument itself so a narrower list fails the build.
`TestComputePlanHealth_BehindScheduleNoteClampsExpectedStepToTotalSteps` asserts
the rendered note text, which the pre-existing test over the same arithmetic
never did.

### Not changed

Executions can reach `partially_completed`, which the score counts as neither
failed nor canceled, so a plan whose executions consistently partially fail still
scores 100. That is a scoring-policy decision rather than a correctness bug and is
left for a separate call.

## Summary by CodeRabbit

- **New Features**
  - Added per-plan health scores to the Plans list, including colored badges for healthy/medium/poor and a muted “unknown” state when health data is missing or not computable.
  - Health tooltips now explain the contributing health factors driving each plan’s score.
- **Bug Fixes**
  - Improved robustness for invalid health scores (null, non-finite, out-of-range) and ensured the score badge is omitted when health data is absent.
  - Prevented unsafe markup from health factor notes from being rendered in tooltips.

### CODE REVIEW

**Comment by cristim** [comment]
## Adversarial review of the health-score computation

Reviewed this PR specifically for the failure mode a computed score invites: a plausible-looking number that is wrong. Three rounds, each with an independent reviewer on the fix diff. Six confirmed defects, all fixed with regression tests that fail pre-fix. Commits `7701298`, `a9a889f`, `45f351e`.

### Confirmed and fixed

**1. Truncated inputs produced confidently-healthy badges** — `internal/api/handler_plans.go`

The execution counts came from `GetExecutionsByStatuses`, whose `ORDER BY scheduled_date DESC + LIMIT` page is capped at `DefaultListLimit` across **all** plans. Once system-wide execution volume exceeded that page, a plan's own failures fell outside it and it rendered a green "healthy" badge while being unhealthy — and the effect grew monotonically as newer rows pushed older ones out.

Replaced with a new `CountExecutionsByPlanAndStatus` store method: a `GROUP BY plan_id, status` aggregate whose result is bounded by plans x statuses rather than execution volume, so the counts are exact and need no limit. A test pins that the capped page is never consulted again.

**2. Fabricated perfect score on a store failure** — `internal/api/handler_plans.go`

When the executions fetch failed, every plan was assigned `HealthScore: 100`. That is a fabricated number on a value operators make purchasing decisions from, and indistinguishable from a measured one. `HealthScore` is now `*int`: it stays nil, serializes as an explicit `null`, and the UI renders a neutral "Health: unknown" badge. The request still succeeds. `null` (uncomputable) and an absent key (older API during a partial deploy) are now distinct states with distinct rendering.

**3. Penalty derived from a missing input** — `internal/api/plan_health.go`

`scheduleFactor` measured elapsed time from `RampSchedule.StartDate` with no zero-value guard, so a plan with no start date was reported ~740000 days behind schedule and docked a real -20. The regression test fails pre-fix with `expected step 15250 by now`. `RampSchedule.GetNextPurchaseDate` already guarded the same field the same way.

**4. A single NULL `plan_id` would have switched the whole feature off** — `internal/config/store_postgres.go`

`purchase_executions.plan_id` has been nullable since migration 000033 (direct-execute purchases have no originating plan; deleting a plan SET NULLs its executions), so the new `GROUP BY` emits a NULL group as soon as one such row is failed or canceled. Scanning that into a plain `string` returns `cannot scan NULL into *string`, which propagated out and blanked the score for *every* plan. Now excluded in SQL plus a `sql.NullString` scan guard, matching `scanExecutionRows` on the same column.

**5. Counting window used the wrong column, making the tooltip false** — `internal/config/store_postgres.go`

A plan's executions are created up front for the whole ramp, so a pending row carries a `scheduled_date` months in the **future**, and `CancelExecutionAtomic` leaves that date untouched when it flips the row to canceled. Bounding on `scheduled_date` therefore matched every future-dated cancellation and described a purchase cancelled today as one of "N canceled execution(s) in the last 30 days" while pointing at next year's date. The window now uses `updated_at` — stamped explicitly by `CancelExecutionAtomic` / `TransitionExecutionStatus` and backstopped by the `update_purchase_executions_updated_at` trigger — so it means when the row entered the status being counted.

**6. Unbounded counts, then a window justified by a claim that did not hold** — `internal/api/plan_health.go`

`CleanupOldExecutions` purges terminal rows but never `failed` ones, so a plan that failed four times once would have been pinned at -40 for the rest of its life with no number of clean runs able to clear it. An intermediate fix bounded the count at 90 days, justified as matching `retention.execution_history_days` — but that setting is read by **nothing** at runtime and the only `CleanupOldExecutions` caller hardcoded 30, so the window matched no retention at all and the two factors were counted over different spans. Both now use the existing `config.DefaultExecutionTTLDays`, and the cleanup caller uses it instead of its own literal, so they cannot drift apart. Every factor note names its window.

### Checked and found clean (not changed)

- **Weights and thresholds**: every penalty, cap, and band boundary was already a named constant with a rationale, not an inline literal. No invented ratios.
- **Division by zero / NaN**: `scheduleFactor` is the only place `StepIntervalDays` is a divisor and it already guarded `<= 0`. Score arithmetic is integer and clamped to [0, 100].
- **XSS**: every API-supplied value reaching `innerHTML` goes through `escapeHtmlAttr`; the existing test asserts on the parsed DOM rather than the HTML string, which is the right shape.
- **Accessibility**: the badge carries a text label ("Health: 72"), so colour is not the sole channel.
- **Savings math**: this PR computes no percentages and touches no `EstimatedSavings` path, so the amortized-upfront and snapshot-averaging concerns do not apply here.
- **Deep-copy**: no getter returns nested mutable state.

### One thing worth a human decision (not changed)

A completed plan short-circuits to a perfect 100 regardless of how many failures it accumulated on the way. That is documented and deliberate in the original commit, and defensible (end-of-life noise is not actionable), so I left it. Flagging it because it means a badge can read 100 for a plan whose history was rocky.

Separately, and pre-existing rather than introduced here: `expireStaleExecutions` transitions stale rows to `"expired"`, which is not in `planHealthExecutionStatuses`, so an approval window silently lapsing never affects a plan's health. Worth a follow-up issue if that is not intended.

### Verification

Root module: `go build ./...` clean, `go test ./...` all 25 packages pass, `golangci-lint` at the CI-pinned **v2.10.1** reports **0 issues** (exit 0), `gocyclo -over 10` clean. Frontend: `npm run lint` clean, `tsc --noEmit` clean, `npm test` 2736 passing, `npm run build` clean. `plans.ts` coverage moved 87.88 -> 87.95 statements / 71.86 -> 72.05 branch.

Note for anyone reproducing locally: `utils`, `approval-details` and `riexchange` fail under a non-en_US system locale (`$1.000` vs `$1,000`) on `main` as well. Unrelated to this PR; they pass under `LC_ALL=en_US.UTF-8`, which is what CI uses.

@coderabbitai full review

**Comment by cristim** [comment]
The previous full-review pass returned without posting findings (included-review limit). Re-requesting now that the window has reset.

@coderabbitai full review

**Comment by cristim** [comment]
Rebased onto `main` (`7b18bfdef`, #1529) to inherit the grpc `v1.82.1` bump that clears `GO-2026-6061`. Head is now `8c448bac0`; all CI checks are green and merge state is CLEAN.

Also corrected the PR body, which had drifted from the code: the scoring path now uses a new store method `CountExecutionsByPlanAndStatus(ctx, statuses, since)` (SQL `GROUP BY plan_id, status`, `plan_id IS NOT NULL`, windowed on `updated_at` rather than `scheduled_date`), and a failed fetch leaves `health_score` nil so the badge renders an explicit unknown state rather than defaulting to a perfect score.

Note for the reviewer: no CodeRabbit review has ever been submitted on this PR (the earlier attempts were rate-limited), so please treat this as a first full pass over the whole diff rather than an incremental one.

@coderabbitai full review

**Comment by cristim** [comment]
Rate-limit window has reset; re-requesting the full review.

@coderabbitai full review

File: frontend/src/plans.ts [hunk 1]
```diff
@@ -861,6 +861,66 @@ interface BackendPlan {
...
+// HTML attribute (title) via innerHTML (feedback_innerhtml_xss: all
+// API-sourced values must be escaped here even though the backend generates
+// these notes itself).
+function healthBadgeHtml(plan: BackendPlan): string {
+  if (plan.health_score === undefined) return '';
+  // The typeof/isFinite check is defensive: the type says number, but this
+  // value comes off the wire and is interpolated into innerHTML below.
+  if (plan.health_score === null
+      || typeof plan.health_score !== 'number'
+      || !Number.isFinite(plan.health_score)) {
```
[the remark this answers is not in the document]
Comment (cristim) [reply frontend/src/plans.ts]: Agreed and fixed in b9932a4. An out-of-range score is exactly as untrustworthy as a NaN, and the badge already routed NaN/Infinity to "unknown", so range now joins that guard rather than being banded. Used named constants (`HEALTH_SCORE_MIN`/`HEALTH_SCORE_MAX`) pointing back at the openapi `minimum: 0, maximum: 100` rather than inline literals, so the guard and the schema stay tied together.

Added regression coverage for -1, 101 and 1000; verified all three fail against the previous implementation.

**Comment by cristim** [comment]
## CodeRabbit pass 1 addressed (`b9932a4`)

Both findings were valid; both fixed, neither dismissed.

**Actionable — out-of-range scores rendered as valid badges** (`frontend/src/plans.ts`)

Fixed. A finite value outside the 0-100 the schema declares (`openapi.yaml`: `minimum: 0, maximum: 100`) was banded like any other number, so `-1` painted red and `101` painted green as though both had been measured. That is the same failure this PR's other fixes exist to prevent: a plausible-looking badge built from a value nothing vouched for. Range now joins the existing NaN/Infinity guard and routes to the "unknown" badge.

I used named constants (`HEALTH_SCORE_MIN`/`HEALTH_SCORE_MAX`) referencing the schema rather than inline `0`/`100`, so the guard and the contract stay tied together.

**Nitpick — band boundaries uncovered** (`frontend/src/__tests__/plans.test.ts`)

Fixed, and it was a real gap rather than a style point. The existing cases sampled 90/65/30, all mid-band, so flipping either comparison from `>=` to `>` would have silently reclassified every plan sitting exactly on a threshold without failing anything.

Added a table-driven case pinning 80 -> `badge-success`, 79 -> `badge-warning`, 50 -> `badge-warning`, 49 -> `badge-danger`.

### Verification

Both fixes confirmed to fail pre-fix rather than just pass post-fix:

- the three out-of-range cases (-1, 101, 1000) fail against the previous implementation;
- the 80-boundary case fails when the good-band comparison is flipped to `>`, which is precisely the regression the nitpick described.

Frontend: `npm run lint` clean, `tsc --noEmit` clean, `npm test` 2743 passing (up from 2736), `npm run build` clean. `plans.ts` coverage 87.95 -> 87.97 statements, 72.05 -> 72.15 branch. Backend untouched by this commit; `go build ./...` and `go test ./internal/api/...` still clean.

@coderabbitai review

**Comment by cristim** [comment]
Requesting one full pass over the final head (`b9932a4`) rather than relying on the incremental result, so the clean signal covers all four commits rather than only the last diff.

@coderabbitai full review

**Comment by cristim** [comment]
Included-review window should have reset. Requesting the confirmation pass over final head `b9932a4`; the previous full-review attempt returned the fair-usage notice without posting findings, so that result was not a reliable clean.

@coderabbitai full review

**Comment by cristim** [comment]
Requesting a full review (not incremental) now that a window may be open. Triggers are serialised centrally across the queue because CodeRabbit is under an account-level adaptive throttle and parallel pings hold it in backoff.

@coderabbitai full review

**Comment by cristim** [comment]
Requesting a full review (not incremental) now that a window may be open. Triggers are serialised centrally across the queue because CodeRabbit is under an account-level adaptive throttle and parallel pings hold it in backoff.

@coderabbitai full review

**Comment by cristim** [comment]
Requesting a full review (not incremental) now that a window may be open. Triggers are serialised centrally across the queue because CodeRabbit is under an account-level adaptive throttle and parallel pings hold it in backoff.

@coderabbitai full review

**Comment by cristim** [comment]
Requesting a full review (not incremental) now that a window may be open. Triggers are serialised centrally across the queue because CodeRabbit is under an account-level adaptive throttle and parallel pings hold it in backoff.

@coderabbitai full review

**Comment by cristim** [comment]
## Addressed in d69ed8fd

### Fixed: retention purged canceled rows inside the health window (CR thread on `internal/server/handler.go:294`)

Confirmed and fixed, plus a sibling path the original report did not cover.

`CleanupOldExecutions` retained canceled rows on `scheduled_date` while `CountExecutionsByPlanAndStatus` counts them on `updated_at`, and `CancelExecutionAtomic` stamps `updated_at` while leaving `scheduled_date` untouched. A plan's executions are created up front and `parseCreatePurchasesRequest` validates only the date *format*, not that the start date is in the future, so a purchase canceled today can carry a `scheduled_date` from 40 days ago. The next sweep deleted it while the health score was still counting it, so the plan's badge silently gained up to 20 points overnight with nothing having changed about the plan.

The `expires_at` branch reintroduced the same purge through a sibling path, which is why the fix is three branches rather than the two in the proposed diff: `expires_at` is fixed when the approval window opens and is never rewritten on cancel, and no transition code writes an "expired" status, so a `notified` row whose token expired 40 days ago and that an operator cancels today was purged by that branch the same day. Fixing only the first branch would have left the defect reachable.

The predicate is now three fully parenthesized branches:

- `completed` retains on `scheduled_date` (unchanged intent)
- the canceled spellings retain on `updated_at`
- `expires_at` reaping now skips canceled rows, so their own `updated_at` window governs them alone

Retention is unchanged for every other status, and a genuinely old canceled row still has an old `updated_at`, so nothing accumulates indefinitely. `status` is `NOT NULL` (migration 000001), so the `NOT IN` on branch 3 is not exposed to three-valued logic.

### Fixed: `behind_schedule` tooltip quoted a step the plan does not have

`expectedStep := daysSinceStart / ramp.StepIntervalDays` was never clamped to `TotalSteps`. Elapsed time keeps growing past the ramp's full duration, so a 4-step weekly ramp started 90 days ago rendered "on step 2, expected step 12 by now". The -20 penalty was always correct; the number the operator was asked to reconcile against was fabricated by the arithmetic. The comparison and the note are now clamped together, so they cannot desync.

### Regression guards, each confirmed failing before the fix

- **`TestPostgresStoreDB_CleanupOldExecutions_RetainsCanceledInsideHealthWindow`** (integration, real PostgreSQL). Asserts exactly which of 7 seeded rows survive a 30-day sweep. Pre-fix the database deleted all three rows that must be retained: canceled-today with a 40-day-old `scheduled_date`, the same row with an expired token, and the legacy spelling. It also asserts a genuinely old canceled row (`updated_at` 40 days ago) IS still deleted, so the fix cannot pass by simply disabling retention, and that `RowsAffected` matches the rows actually purged. Runs against a real database rather than pgxmock because the defect is in what the predicate *means*, not how it is spelled.
- **`TestPGXMock_CleanupOldExecutions_RetainsCanceledOnUpdatedAt`** pins each branch's retention column in the fast unit suite, one subtest per branch so a failure names the branch that regressed.
- **`TestComputePlanHealth_BehindScheduleNoteClampsExpectedStepToTotalSteps`** asserts the rendered note text. `TestComputePlanHealth_ScoreClampsAtZeroWhenPenaltiesStack` already exercised this exact arithmetic but asserted only codes and the clamped score, which is why nothing caught it; it now asserts the note too.

### Gates

`go build ./...` 0 | `go vet ./internal/...` 0 | `go vet -tags=integration ./...` 0 | `go test ./internal/...` 0 | `go test -race -tags=integration` (new test) 0 | gocyclo `-over 10` (pre-commit form) 0 | golangci-lint **v2.10.1**, the version pinned in `ci.yml`, `./internal/...` 0 issues | full pre-commit suite passed on commit. Frontend untouched.

### Not changed (flagged for a product call, not a defect)

Executions can reach `partially_completed`, which the score counts as neither failed nor canceled, so a plan whose executions consistently partially fail still scores 100. Whether a partial failure should carry a penalty is a scoring-policy decision rather than a correctness bug, so scoring behaviour is left as-is here.

@coderabbitai review

**Comment by cristim** [comment]
Re-requesting review: the latest commit on this PR postdates your last review, so that review does not cover the current HEAD.

Fixes landed since then were verified locally (build, tests, vet, gocyclo, and golangci-lint at the CI-pinned version), and CI on this PR is green across all workflows. Please review the current HEAD.

@coderabbitai review

File: internal/server/handler.go [hunk 2]
```diff
@@ -286,8 +287,11 @@ func (app *Application) handleCleanupExpiredRecords(ctx context.Context) (map[st
 
 	// Clean up old execution records (30+ days)
 	if app.Config != nil {
-		const retentionDays = 30
-		deleted, err := app.Config.CleanupOldExecutions(ctx, retentionDays)
+		// Shared with the plan health score's lookback window
+		// (internal/api/plan_health.go): that score counts failed/canceled
+		// executions over exactly the period this sweep guarantees is
+		// retained, so the two must never drift apart.
+		deleted, err := app.Config.CleanupOldExecutions(ctx, config.DefaultExecutionTTLDays)
```
[the remark this answers is not in the document]
Comment (cristim) [reply internal/server/handler.go:294]: Confirmed and fixed in d69ed8fd, with a follow-up in 9166bc5d. Leaving this thread open for you to re-review rather than resolving it.

The predicate is now split as you proposed, with the branches fully parenthesized so the AND/OR precedence cannot shift:

```sql
WHERE (      status = 'completed'
         AND scheduled_date < NOW() - INTERVAL '1 day' * $1 )
   OR (      status IN ('cancelled', 'canceled')
         AND updated_at     < NOW() - INTERVAL '1 day' * $1 )
   OR (      status <> ALL($2)
         AND expires_at IS NOT NULL
         AND expires_at     < NOW() - INTERVAL '1 day' * $1 )
```

Two things go beyond the proposed diff, both found while verifying it.

**The `expires_at` branch reintroduced the same purge.** `expires_at` is written once at insert and never rewritten when a row later reaches a terminal status, so a row canceled *today* can carry a long-lapsed `expires_at` and be deleted by that branch on the same sweep, while the score is still counting it. Splitting only the first branch would have left the defect reachable, so branch 3 now skips the statuses the score counts.

**It is not only `canceled`.** The score also counts `failed`, at -10 each up to -40, which is twice canceled's worst case, so `failed` had the larger version of the same bug. Rather than maintain a second literal status list that can drift from the score's, the exclusion is parameterized as `$2` from `config.HealthScoredExecutionStatuses`, the same slice `plan_health.go` passes to `CountExecutionsByPlanAndStatus`. Adding a status to the score now extends the sweep's exclusion in the same edit.

One correction to my own earlier claim, for the record: the `expires_at` arm is **not** reachable on rows current code writes. That column has exactly one writer, `SavePurchaseExecution` via `timeFromTTL(execution.TTL)`, and no production path assigns `PurchaseExecution.TTL`, so it is NULL on every row created today and the branch is inert for them. The exclusion is defence in depth for legacy imported rows. Whether any such row carries a non-NULL `expires_at` needs a production query; it cannot be settled from code.

**Regression coverage**, as you asked for. The behavioural proof is an integration test against a real PostgreSQL rather than pgxmock, because the defect is in what the predicate *means*, not how it is spelled:

```
go test -tags=integration ./internal/config/ \
  -run TestPostgresStoreDB_CleanupOldExecutions_RetainsCanceledInsideHealthWindow
```

The `-tags=integration` is load-bearing: without it the file is excluded by its build tag and the run reports "no tests to run" and exits 0, which looks like a pass.

It seeds nine rows across the status x timestamp matrix and asserts exactly which survive a 30-day sweep. Confirmed failing before each fix: pre-`d69ed8fd` the database deleted all three rows that must be retained (canceled-today with a 40-day-old `scheduled_date`, the same row with a lapsed `expires_at`, and the legacy spelling); pre-`9166bc5d` it deleted the `failed`-today row, which the suite had no case for at all. It also asserts that a genuinely old canceled row IS still deleted and that `RowsAffected` matches, so neither fix can pass by quietly disabling retention.

A user-visible consequence worth flagging on this thread: a canceled row with a **future** `scheduled_date` was previously retained forever, since `scheduled_date < NOW() - retention` never became true for it. It now leaves History 30 days after the cancellation instead of 30 days after the date it was scheduled for. That direction closes an unbounded-retention leak and matches what the health window already assumed, but it is a change in behaviour rather than a pure bug fix.

**Comment by cristim** [comment]
## Round 2 — 9166bc5d

Independent verification of `d69ed8fd` reproduced the pre-fix failure and probed the full status x timestamp matrix against a real PostgreSQL. Both original fixes held in every cell. It surfaced one more instance of the same defect plus three inaccurate claims, all addressed here in one commit.

**The `expires_at` exclusion was one status too narrow.** It skipped only the canceled spellings, but the health score also counts `failed`, at -10 each up to -40, which is twice canceled's worst case. Since `expires_at` is written once at insert and never rewritten when a row later reaches a terminal status, a row that failed *today* with a lapsed `expires_at` was deleted while the score was still counting it. Same overnight score jump, larger magnitude.

The exclusion is now parameterized as `$2` from `config.HealthScoredExecutionStatuses`, the same slice `plan_health.go` passes to `CountExecutionsByPlanAndStatus`, so adding a status to the score extends the sweep's exclusion in the same edit rather than drifting from a second hand-maintained list. `failed` still has no cleanup branch of its own, which is what `plan_health.go`'s lookback comment already claimed and the score depends on.

**Corrected an overstatement in my own earlier description.** I presented the `expires_at` arm as a live defect. It is not reachable on rows current code writes: that column has exactly one writer, `SavePurchaseExecution` via `timeFromTTL(execution.TTL)`, and no production path assigns `PurchaseExecution.TTL`, so it is NULL on every row created today and the branch is inert for them. The exclusion is defence in depth for legacy imported rows. Whether any such row carries a non-NULL `expires_at` needs a production query and cannot be settled from code; the PR body now says that instead of implying a live bug.

**Two stale doc claims fixed.** `expires_at` is the row's own TTL column, not the approval-token deadline — that is `approval_token_expires_at`, added by migration 000051. And the comment justifying the branch said no code writes an `'expired'` status and the CHECK excludes it; both are false (migration 000089's `purchase_executions_status_check` lists `'expired'`, and `expireStaleExecutionsSweep` writes it). Behaviourally harmless, since an `'expired'` row is unscored and still governed by that branch, but it was the branch's stated justification.

**Behaviour change now disclosed** in the PR body: a canceled row with a *future* `scheduled_date` was previously retained forever, and now leaves History 30 days after cancellation rather than 30 days after its scheduled date.

### Regression guards, both confirmed failing first

The `failed` case is new; the suite had no failed-row case at all, so it stayed green with the defect present. Reverting only the exclusion widening fails exactly `failed-today-with-lapsed-expires-at` plus the `RowsAffected` assertion. A second pgxmock guard pins the `$2` argument itself rather than just the `<> ALL($2)` spelling, so passing a narrower list fails the build.

Verify with the tag — without it the file is excluded by its build tag and the run exits 0 having executed nothing:

```bash
go test -tags=integration ./internal/config/ \
  -run TestPostgresStoreDB_CleanupOldExecutions_RetainsCanceledInsideHealthWindow
```

### Gates

`go build ./...` 0 | `go vet ./internal/...` 0 | `go vet -tags=integration ./...` 0 | `go test ./internal/...` 0 (19 pkgs) | integration test 0, and 0 under `-race` | gocyclo `-over 10` (pre-commit form) 0 | **golangci-lint v2.10.1**, the version pinned in `ci.yml`, `./internal/...` → `0 issues` | full pre-commit suite passed on commit. All 6 CI workflows on the previous commit `d69ed8fd` finished green.

Two existing pgxmock cases were updated for the second query argument.

@coderabbitai full review
